test/merger/test_merger.py

```diff
@@ -172,3 +174,118 @@ def test_no_merge_across_reset() -> None:
     builder2.reset(0)
     expected_qc = builder2.to_circuit()
     modify_circuit_and_check(qc, general_merger.merge_single_qubit_gates, expected_qc)
+
+
+@pytest.mark.parametrize(
+    ("circuit", "expected_result"),
+    [
+        (
+            CircuitBuilder(2).H(0).barrier(0).H(1).barrier(1).H(0).Rx(0, Float(math.pi / 3)).barrier(0).to_circuit(),
+            """version 3.0
+
+qubit[2] q
+
+H q[0]
+H q[1]
+barrier q[0]
+barrier q[1]
+Anonymous gate: BlochSphereRotation(Qubit[0], axis=[ 0.65465 -0.37796  0.65465], angle=-2.41886, phase=1.5708)
```

Comment:
Barrier 1 is merged with the first barrier 0 because it is more efficient to do so. As you said, it does not have anything to do with the last bit, so it will try to be scheduled earlier in the process than later. This is the strategy we have been implicitly using for all our other merges.

Reply:
@juanboschero: I apologize for going back on the decision on optimizing the instructions with respect to the barriers. Optimization/reordering should be part of a separate pass, so I want to keep it disentangled here as well. So for now we stick with the less efficient `merge_single_qubit_gates`.